src/ucp/proto/proto_single.c

```diff
@@ -10,13 +10,89 @@
 
 #include "proto_single.h"
 #include "proto_common.h"
+#include "proto_common.inl"
 #include "proto_init.h"
 #include "proto_debug.h"
 
 #include <ucs/debug/assert.h>
 #include <ucs/debug/log.h>
 #include <ucs/sys/math.h>
 
+static double
+ucp_proto_single_get_bandwidth(const ucp_proto_common_init_params_t *params,
+                               ucp_lane_index_t lane)
+{
+    ucp_proto_common_tl_perf_t tl_perf;
+    ucp_proto_perf_node_t *perf_node;
+    ucs_status_t status;
+
+    status = ucp_proto_common_get_lane_perf(params, lane, &tl_perf, &perf_node);
+    if (status != UCS_OK) {
+        return 0;
+    }
+
+    ucp_proto_perf_node_deref(&perf_node);
+    return tl_perf.bandwidth;
+}
+
+static void
+ucp_proto_single_update_lane(const ucp_proto_single_init_params_t *params,
+                             ucp_lane_index_t *lane_p)
+{
+    const ucp_proto_common_init_params_t *common_params = &params->super;
+    const ucp_proto_init_params_t *init_params = &common_params->super;
+    const ucp_context_h context                = init_params->worker->context;
+    double bandwidth;
+    ucp_lane_index_t lanes[UCP_PROTO_MAX_LANES];
+    ucs_sys_device_t sys_devs[UCP_PROTO_MAX_LANES];
+    ucp_lane_index_t num_lanes, num_identical_devs, i, lane, j;
+    ucs_sys_device_t sys_dev;
+
+    if (context->config.ext.proto_use_single_net_device &&
+        (context->config.node_local_id != 0)) {
```

Comment:
why? Should we check for some specific value meaning uninitalized?

Reply:
just a small optimization, since lane[0] will be chosen for the zero id anyway.

Reply:
can you pls add comment then?